### ISSUE TITLE
add/delete many-to-many Problem

### ISSUE THREAD

**Issue by MarcDrexler** [TrackableEntities/trackable-entities#131]
Hello Tony,

I don't know if I am doing something wrong or if there is a problem with the M-M relationship.
I tested M-M relationships between employee and territory.

I wanted to remove a single territory from the Territories Collection. But if I do that, the TrackingState of employee is also set to Deleted. This will cause the context to delete my employee, which I dont want :-)

```
var employee = employeeService.GetEmployeeAsync(1).Result;
var changeTracker = new ChangeTrackingCollection<Employee>(employee);
employee.Territories.Remove(employee.Territories.First());
```

I have the same problem when I add a territory to the Territories Collection. The TrackingState of employee is then set to Added.

I only want to add or remove a territory, but I cannot figure out how to do that.

**Comment by tonysneed** [comment]
@MarcDrexler,

Thanks for your question.  First, make sure you have installed the latest version of the client NuGet package.  Adding and removing a territory should leave the employee marked as UnChanged.  See the unit test: [Existing_Employee_With_Territories_Should_Have_Territories_Marked](https://github.com/TrackableEntities/trackable-entities/blob/release-2.5.4/Source/Tests/TrackableEntities.Client.Tests/ChangeTrackingCollectionWithChildrenTests.cs#L593).

So I would first suggest you try to replicate this test using your own code, and to see the different from what you were trying.  That may shed some light.  If that doesn't help, then perhaps you could try setting up a repro with a unit test that shows what you were experiencing.  You can do this either by forking the trackable-entities repo, then adding a unit test there, or by creating your own repo with the test, which I can fork in order to see the behavior.

Thanks!

**Comment by MarcDrexler** [comment]
@tonysneed, thanks for the reply and sorry for my late answer. Finally I got some time to setup a test that demonstrates the behaviour.
The unit test you mentioned is working fine. But when I use the wcf sample project, I encounter the problem.

I forked your repository and modified the solution WcfSample.VS2013:
https://github.com/MarcDrexler/trackable-entities/blob/develop/Samples/VS2013/WcfSample/WcfSample.VS2013.sln

I added one more service that serves territories and employees: EmployeeService
And [this unit test](https://github.com/MarcDrexler/trackable-entities/blob/develop/Samples/VS2013/WcfSample/WcfSample.Tests/EmployeeTests.cs#L35) demonstrates my problem.

I am using the latest nuget packages.

**Comment by tonysneed** [comment]
Thanks @MarcDrexler, I'll be sure to have a look at this.

**Comment by tonysneed** [comment]
@MarcDrexler, I can see the behavior with your sample and unit test.  It appears to have something to do with recursion and object visitation.  When I set the Employees property of the deleted territory to null, then the problem goes away.  I only see this issue with the WCF sample and not with the Web API sample.

@azabluda, Would you possibly be able to take a look at this?

**Comment by azabluda** [comment]
The reason why it works in Tony's case, but does not in Marc's is this line https://github.com/TrackableEntities/trackable-entities/commit/901b728148b8f8e5e2f580b3b5df1271a6aea4ad#diff-9cf5d267b1b896dfc9afc018f02a1707R100. Apparently [it is still missing](https://github.com/TrackableEntities/trackable-entities/blob/release-2.5.4/Samples/VS2013/WcfSample/WcfSample.Shared.Entities.Net45/Models/Employee.cs#L110) in WcfSample's entities. Either the code generation template was not updated to reflect the a.m. revision, or it was not re-applied to the sample project.

**Comment by tonysneed** [comment]
Great catch Alex - that explains it!

The entity code gen templates are now deployed in NuGet packages.  I'll update the samples to use those.

**Comment by MarcDrexler** [comment]
Thank you both!

I think the code gen templates in the currently released NuGet packages are also not up to date.

I have just created a new project (shared entities, .net 4.5) with the help of the VS2015 extension 2.5.2.1.
This project contains the NuGet package "TrackableEntities.CodeTemplates.Shared.Net45" v2.5.0.

I follow these steps written in Readme.txt:
> To get started, just right-click on the project produced by the
> template, select Add New Item, Data, ADO.NET Entity Data Model.
> Select or create a data connection. The tool will generate Code First
> entities based on the T4 templates located in the project.
> 

I can see generated entities, but the neccesary lines for M-M relations are still not included.

**Comment by tonysneed** [comment]
I'll have a look.

**Comment by MarcDrexler** [comment]
Hi Tony,
i'm still trying to generate shared entities using your latest code gen templates, but I am unable to find them :-/
Can you point me to the latest code gen sources in your repository?

**Comment by tonysneed** [comment]
Ah yes, thanks for your patience.  The source code for the templates is here: https://github.com/TrackableEntities/trackable-entities/tree/develop/Extensions/Templates

And the NuGet packages for the templates can be found here:
https://github.com/TrackableEntities/trackable-entities/tree/develop/NuGet/Packages/TrackableEntities.CodeTemplates

Have you tried installing the appropriate NuGet package and then generating entities?

It's possible that the NuGet package is up-to-date, but that the project template is not.

**Comment by MarcDrexler** [comment]
Hi Tony,

I understand that it is possible that the project template is not up-to-date. But this is ok for me. I tried generating the entities with the latest NuGet package.
However I am still unable to see [these](https://github.com/TrackableEntities/trackable-entities/commit/901b728148b8f8e5e2f580b3b5df1271a6aea4ad#diff-9cf5d267b1b896dfc9afc018f02a1707R100) necessary lines mentioned by azabluda.

Obviously, something I am doing must be wrong.
I have recorded a [short video](https://dl.dropboxusercontent.com/u/5746496/TrackableEntities.mp4) so you can have a look how I am generating my entities.

**Comment by tonysneed** [comment]
Thanks! Appreciate the reminder also, which helps nudge the issue to the top of my queue. :)

**Comment by tonysneed** [comment]
Corrected the problem with PR #135.  Simply update the CodeTemplate NuGet packages to v. 2.5.2, then delete and regenerate entities for M-M relations.

**Comment by MarcDrexler** [comment]
Great, it is working now!
Thanks so much for your time and help! :-)
